Block Directory: Store refactor

ryelle opened the pull request:
## Description
- Pulls in some of the work from #20001 (can be rebased out once that merges)
- Cleans up jsdoc descriptions & formatting
- Switches to using `data-controls`, which cuts down `controls` to just the `LOAD_ASSETS` action
- Removes uninstall code, so we don't accidentally end up with duplicate code when we re-implement it
- Refactor the install & download actions to be a single `installBlockType` action generator, which walks through each step. This avoids needing to add extra logic into the `withDispatch` for error/success callbacks
- Updates the notice action/reducer to store the error message itself, which will let us show the user more descriptive error messages.

## How has this been tested?
I've added tests for actions, reducers, selectors, and to some extent the controls helpers. I also did some quick browser-testing— this removes any error notices (because of the switch in 20001), so it's not ready to merge until that's done.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

ryelle committed:
Pass along action parameter

ryelle committed:
Fix up comments

ryelle committed:
Use shared package `data-controls`

ryelle committed:
Remove `uninstallBlock`, `removeInstalledBlockType`

ryelle committed:
Update controls to return promises

Includes very basic testing, since I'm having trouble with both promises and document mocking

ryelle committed:
Update response format from API

ryelle committed:
Refactor install & download into one single, multi-step action generator

This walks through each step, and removes the need for onError & onSuccess

ryelle committed:
Switch to new action generator for installation

ryelle committed:
Update Notice reducer to store the notice text

This will let us return more descriptive error messages for each kind of error, instead of generic install/download failed messages

ryelle committed:
Bring `onSelect` functionality back

ryelle committed:
Fix notice description

ZebulanStanphill commented on packages/block-directory/src/components/downloadable-blocks-list/index.js:
```diff
@@ -59,78 +52,13 @@ function DownloadableBlocksList( {
-		const { removeBlocks } = dispatch( 'core/block-editor' );
-		const { onSelect } = props;
-
+	withDispatch( ( dispatch, { onSelect } ) => {
```
While you're at it, you could refactor the component to just use the `useDispatch` hook instead of the `withDispatch` HOC.

ryelle replied:
Good idea, thanks!

ryelle committed:
Switch to `useDispatch` hook

StevenDufresne commented on packages/block-directory/src/store/controls.js:
```diff
@@ -1,153 +1,71 @@
+		const existing = document.querySelector( `script[src="${ asset }"]` );
+		if ( existing ) {
+			existing.parentNode.removeChild( existing );
+		}
```
Why does this matter?

ryelle replied:
TBH It was in the original code and I didn't remove it. I think it's used when retrying loading assets, so we don't end up with duplicate script tags in the document.

StevenDufresne replied:
right.... to prevent issues during the injection step on retry. ok.

StevenDufresne commented on packages/block-directory/src/store/controls.js:
```diff
@@ -1,153 +1,71 @@
+		link.rel = 'stylesheet';
+		link.href = asset;
+		link.onload = () => resolve( true );
+		link.onerror = () => reject( new Error( 'Error loading script.' ) );
```
```suggestion
		link.onerror = () => reject( new Error( 'Error loading style.' ) );
```

StevenDufresne commented on packages/block-directory/src/store/controls.js:
```diff
@@ -1,153 +1,71 @@
-		return new Promise( ( resolve, reject ) => {
-			if ( Array.isArray( assets ) ) {
-				let scriptsCount = 0;
+		const scripts = assets.map( ( asset ) =>
```
With the API being premature, and the control of block.json files being in the author's hand, its possible that we get non .js/.css files.

ryelle replied:
That's fair— do you think we should do this validation here in the loading code, or back in the API itself?

ryelle replied:
actually, rethinking this— on the basis of "escaping late", I'll add a check in loadScript/loadStyle to verify that at least the extension is valid.

If there ends up being a valid reason to load another file type, we can change this conditional then.

StevenDufresne replied:
Cool.

StevenDufresne reviewed (commented):
Awesome. Thanks for the taking the time. 
I left a few comments on choices the code already makes. They are not necessary for merge, but worth discussing in the context of a refactor.

StevenDufresne commented:
You have this as a draft PR, what else do you plan on refactoring?

ryelle committed:
Update packages/block-directory/src/store/controls.js

Co-authored-by: Steven Dufresne <steve.dufresne@automattic.com>

ryelle committed:
Remove extra export

ryelle committed:
Remove unnecessary resets

ryelle committed:
Update variable name to reflect value

ryelle committed:
Add test to block list item

ryelle committed:
Remove unused props

ryelle committed:
Remove behavior test, update to use snapshot testing

Behavior is tested in DownloadableBlockListItem tests

ryelle committed:
Update notice block to use `useSelect`

ryelle committed:
Update package-lock

ryelle committed:
Check file extension before trying to load scripts & styles

ryelle committed:
Fix class name

ryelle commented:
> You have this as a draft PR, what else do you plan on refactoring?

I wanted to rebase it after #20001 was merged. Now that's done, so this is ready for review. I also refactored some of the components & tests (like `DownloadableBlockNotice`, since I introduced the `getErrorNoticeForBlock` selector).

ryelle committed:
Bring back `setIsInstalling`

StevenDufresne commented on packages/block-directory/src/store/actions.js:
```diff
@@ -45,88 +50,44 @@ export function setInstallBlocksPermission( hasPermission ) {
 		} );
-		if ( response.success === false ) {
+		if ( response.success !== true ) {
 			throw new Error( response.errorMessage );
```
I tested locally, if the length of this error message is `< 1`, no error message will be displayed.

StevenDufresne replied:
I wonder if we should consider a default error message?

ryelle replied:
Added the oh-so-descriptive "An error occurred." as a default. Hopefully we'll have error content for most detectable errors, so no one will see this 🙂

TimothyBJacobs replied:
If an error occurs here it'd be a proper REST API error response. AFAIK, that throws. I don't think this code path could get executed. Am I missing something?

ryelle replied:
Yeah, this would probably never be executed (and if it is, errorMessage wouldn't exist) — but it could if the response from the server was malformed somehow, or the API endpoint could change to return failure info. I've updated it to show a default message, and we can look at removing it if needed when the API settles some.

StevenDufresne commented:
Just nothing here that the API has some issues that cause `500` errors. 
This should be corrected in:  
https://github.com/WordPress/gutenberg/pull/22454 (or a related PR, @tellyworth has one going).

This is the result:

![Error](https://d.pr/i/sUUrAP.png)

TimothyBJacobs commented:
@StevenDufresne do you know what API route was called that ended up triggering that error? Want to make sure it gets corrected.

StevenDufresne reviewed (commented):
> @StevenDufresne do you know what API route was called that ended up triggering that error? Want to make sure it gets corrected.

This was the result of the plugin already being installed -> `/install`. 

I'm certain it's the result of `new` missing from `WP_Error` statements.

ryelle committed:
Add i18n & punctuation to error messages

ryelle committed:
Add a default error message

ryelle committed:
Return a success/failure indicator for installation

This prevents an error where `onSelect` tries to run on a block that fails to install

ryelle commented on packages/block-directory/src/store/actions.js:
```diff
@@ -45,88 +50,44 @@ export function setInstallBlocksPermission( hasPermission ) {
 	try {
-		if ( ! item.assets.length ) {
+		if ( ! Array.isArray( assets ) || ! assets.length ) {
 			throw new Error( 'Block has no assets' );
```
Done— I also added the i18n wrapper, since these are shown to users now.

StevenDufresne commented:
Looks good.

ryelle committed:
Remove unused state property, update tests

ryelle committed:
Add selector tests for `getDownloadableBlocks`

ryelle committed:
Update mocked API response

ryelle committed:
Add default message for installation failures

StevenDufresne reviewed (approved):
I'm good with the changes.
